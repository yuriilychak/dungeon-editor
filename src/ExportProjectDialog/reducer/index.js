import STATE from "../state";
import StaticData from "../../data/ExportProjectDialogData";

export const initialState = {
    staticData: StaticData,
    isPopupOpen: false
};

const actionHandlers = {
    [STATE.CHANGE_ACTIVITY]: (state, action) => {
        return {
            ...state,
            isPopupOpen: action.payload
        }
    }
};

export default function topMenuReducer(state = initialState, action) {
    const actionHandler = actionHandlers[action.type];
    return actionHandler ? actionHandler(state, action) : state;
}
