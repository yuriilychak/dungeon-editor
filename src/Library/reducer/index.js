import StaticData from "../data/index";
import STATE from "../state";

/**
 * @typedef {Object}
 * @name LibraryData
 * @property {number} id
 * @property {string} locale
 */

/**
 * @typedef {Object}
 * @name LibraryState
 * @property {Array<LibraryData>} tabs
 * @property {Array<Object>} files
 */

/**
 * @type {LibraryState}
 */

export const initialState = {
    ...StaticData,
    files: [[], [], [], [], []]
};

/**
 * @type {Object.<string, function(LibraryState, ActionData): LibraryState>}
 */

const actionHandlers = {
    [STATE.ADD_FILE]: (state, action) => {
        const files = state.files.slice(0);
        const {sectionId, data} = action.payload;
        const file = {
            title: data.name,
            id: data.id,
            hasPreview: false,
            isDirectory: false
        };
        files[sectionId] = [...files[sectionId], file];
        return {
            ...state,
            files
        }
    },
    [STATE.REMOVE_FILE]: (state, action) => {
        const files = state.files.slice(0);
        const index = action.payload.sectionId;
        files[index] = files[index].filter(file => file.id !== action.payload.id);
        return {
            ...state,
            files
        }
    },
    [STATE.UPDATE_TREE]: (state, action) => {
        const files = state.files.slice(0);
        const {fileTree, sectionId} = action.payload;
        files[sectionId] = fileTree;
        return {...state, files};
    },
    [STATE.ADD_DIRECTORY]: (state, action) => {
        const files = state.files.slice(0);
        const {sectionId, dirData} = action.payload;
        const fileTree = files[sectionId].slice(0);
        const parentId = dirData.parentId;
        const result = {
            title: dirData.name,
            id: dirData.id,
            hasPreview: false,
            isDirectory: true
        };

        if (parentId === -1) {
            fileTree.push(result);
        }
        else {
            const parent = findDir(fileTree, parentId);

            if (!parent) {
                return state;
            }

            if (!parent.children) {
                parent.children = [];
            }

            parent.children.push(result);
        }

        files[sectionId] = fileTree;

        return {...state, files};
    },
    [STATE.CLEAR]: state => ({...state, files: [[], [], [], [], []]})
};

function findDir(data, id) {
    const dirs = data.filter(element => element.isDirectory);
    let result = dirs.find( dir => dir.id === id);

    if (result) {
        return result;
    }

    dirs.forEach( dir => {
        if (!dir.children || result) {
            return;
        }

        result = findDir(dir.children, id);
    });

    return result;
}

/**
 * @param {LibraryState} state
 * @param {ActionData} action
 * @return {LibraryState}
 */

export default function topMenuReducer(state = initialState, action) {
    const actionHandler = actionHandlers[action.type];
    return actionHandler ? actionHandler(state, action) : state;
}
