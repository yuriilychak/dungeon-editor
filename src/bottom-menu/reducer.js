import StaticData from "./data";
import STATE from "./state";
export const initialState = {
    ...StaticData,
    sectionId: -1,
    fileId: -1,
    tabIndex: 0,
    elements: []
};

const actionHandlers = {
    [STATE.OPEN_ELEMENT]: (state, { fileId, sectionId }) => {
        return {
            ...state,
            fileId,
            sectionId,
            tabIndex: 0
        }
    },
    [STATE.CHANGE_TAB]: (state, tabIndex) => ({ ...state, tabIndex })
};

export default function topMenuReducer(state = initialState, { type, payload}) {
    const actionHandler = actionHandlers[type];
    return actionHandler ? actionHandler(state, payload) : state;
}

