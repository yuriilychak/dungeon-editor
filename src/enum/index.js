export { default as SECTION_ID } from "./section-id";
export { default as EDIT_MODE } from "./edit-mode";
export { default as STAGE_ELEMENT_PROP } from "./stage-element-prop";
export { default as VALUE_FORMAT } from "./value-format";
export { default as FIELD_TYPE } from "./field-type";
export { default as UI_SECTION } from "./ui-section";
export { default as FILE_FORMAT } from "./file-format";
