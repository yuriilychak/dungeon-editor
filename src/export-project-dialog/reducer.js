import STATE from "./state";
import {generateReducerData} from "../helpers";

/**
 * @typedef {Object}
 * @name ProgressData
 * @property {number} progress
 * @property {string} fileName
 * @property {boolean} isComplete
 */

/**
 * @typedef {Object}
 * @name ExportProjectDialogState
 * @property {Object.<string, string>} locale
 * @property {boolean} isPopupOpen
 * @property {?ProgressData} progressData
 */


/**
 * @param {ExportProjectDialogState} state
 * @param {ActionData} action
 * @return {ExportProjectDialogState}
 */

export default generateReducerData(
    {
        isPopupOpen: false,
        progressData: null
    },
    {
        [STATE.CHANGE_ACTIVITY]: (state, isPopupOpen) => ({
            ...state,
            isPopupOpen,
            progressData: null
        }),
        [STATE.PROGRESS]: (state, progressData) => ({ ...state, progressData })
    }
);
