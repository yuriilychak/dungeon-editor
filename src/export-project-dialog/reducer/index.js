import STATE from "../state";
import StaticData from "../data";

/**
 * @typedef {Object}
 * @name ProgressData
 * @property {number} progress
 * @property {string} fileName
 * @property {boolean} isComplete
 */

/**
 * @typedef {Object}
 * @name ExportProjectDialogState
 * @property {Object.<string, string>} locale
 * @property {boolean} isPopupOpen
 * @property {?ProgressData} progressData
 */

/**
 * @type {ExportProjectDialogState}
 */

export const initialState = {
    ...StaticData,
    isPopupOpen: false,
    progressData: null
};

/**
 * @type {Object.<string, function(ExportProjectDialogState, ActionData): ExportProjectDialogState>}
 */

const actionHandlers = {
    [STATE.CHANGE_ACTIVITY]: (state, action) => ({
        ...state,
        isPopupOpen: action.payload,
        progressData: null
    }),
    [STATE.PROGRESS]: (state, action) => ({
        ...state,
        progressData: action.payload
    })

};

/**
 * @param {ExportProjectDialogState} state
 * @param {ActionData} action
 * @return {ExportProjectDialogState}
 */

export default function topMenuReducer(state = initialState, action) {
    const actionHandler = actionHandlers[action.type];
    return actionHandler ? actionHandler(state, action) : state;
}
