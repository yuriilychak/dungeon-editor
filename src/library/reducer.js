import StaticData from "./data";
import STATE from "./state";
import { handleAction } from "../helpers";

/**
 * @typedef {Object}
 * @name LibraryData
 * @property {number} id
 * @property {string} locale
 */

/**
 * @typedef {Object}
 * @name LibraryState
 * @property {Array<LibraryData>} tabs
 * @property {Array<Object>} files
 */

/**
 * @type {LibraryState}
 */

export const initialState = {
    ...StaticData,
    files: [[], [], [], [], [], [], []]
};

/**
 * @type {Object.<string, function(LibraryState, ActionData): LibraryState>}
 */

const actionHandlers = {
    [STATE.ADD_DIRECTORY]: (state, payload) => addElementToTree(payload, true, state),
    [STATE.ADD_FILE]: (state, payload) => addElementToTree(payload, false, state),
    [STATE.REMOVE_FILE]: (state, payload) => {
        const files = state.files.slice(0);
        const { sectionId, id } = payload;
        const fileTree = files[sectionId];
        const element = findElement(fileTree, id);
        const parentId = element.parentId;

        if (parentId === -1) {
            files[sectionId] = fileTree.filter(element => element.id !== id);
        }
        else {
            const parent = findElement(fileTree, parentId);
            parent.children = parent.children.filter(element => element.id !== id);

            files[sectionId] = fileTree.slice(0);
        }

        return {
            ...state,
            files
        }
    },
    [STATE.UPDATE_TREE]: (state, payload) => {
        const files = state.files.slice(0);
        const {fileTree, sectionId} = payload;
        refreshParentIds(fileTree);

        files[sectionId] = fileTree;

        return {...state, files};
    },
    [STATE.RENAME_FILE]: (state, payload) => {
        const files = state.files.slice(0);
        const {id, sectionId, name} = payload;
        const fileTree = files[sectionId];

        const element = findElement(fileTree, id);

        if (!element) {
            return state;
        }

        element.title = name;

        files[sectionId] = fileTree.slice(0);

        return {...state, files};
    },
    [STATE.CLEAR]: state => ({...state, files: [[], [], [], [], [], [], []]})
};

function refreshParentIds(files, parentId = -1) {
    files.forEach(file => {
        file.parentId = parentId;
        if (file.isDirectory && file.children && file.children.length !== 0) {
            refreshParentIds(file.children, file.id);
        }
    })
}

function addElementToTree(payload, isDirectory, state) {
    const files = state.files.slice(0);
    const {sectionId, data} = payload;
    const file = {
        title: data.name,
        id: data.id,
        parentId: data.parentId,
        preview: data.preview,
        isDirectory: isDirectory
    };
    const parentId = file.parentId;
    const fileTree = files[sectionId].slice(0);

    if (parentId === -1) {
        fileTree.push(file);
    }
    else {
        const parent = findElement(fileTree, parentId);

        if (!parent) {
            return state;
        }

        if (!parent.children) {
            parent.children = [];
        }

        parent.children.push(file);
    }

    files[sectionId] = fileTree;

    return {
        ...state,
        files
    };
}

function findElement(data, id) {
    const dirs = data.filter(element => element.isDirectory);
    let result = data.find(element => element.id === id);

    if (result) {
        return result;
    }

    dirs.forEach( dir => {
        if (!dir.children || result) {
            return;
        }

        result = findElement(dir.children, id);
    });

    return result;
}

/**
 * @param {LibraryState} state
 * @param {ActionData} action
 * @return {LibraryState}
 */

export default function topMenuReducer(state = initialState, action) {
    return handleAction(state, actionHandlers, action);
}
