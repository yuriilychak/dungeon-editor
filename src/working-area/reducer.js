import STATE from "./state";
import StaticData from "./data/index.json";
import {WorkingStage} from "../working-stage";

export const initialState = {
    ...StaticData,
    tabs: [],
    selectedTab: 0,
    zoomValue: 1
};

const closeTab = (state, tabIndex) => {
    let {tabs, selectedTab} = state;

    if (tabIndex >= tabs.length) {
        return state;
    }

    if (selectedTab >= tabIndex && selectedTab !== 0) {
        --selectedTab;
    }

    const nextTabs = tabs.filter((tab, index) => index !== tabIndex);

    if (nextTabs.length) {
        const nextTab = nextTabs[selectedTab];

        WorkingStage.showElement(nextTab.sectionId, nextTab.fileId);
    }

    return {
        ...state,
        tabs: nextTabs,
        selectedTab
    };
};

const selectTab = (state, tabIndex) => {
    if (tabIndex === state.selectedTab) {
        return state;
    }

    return {
        ...state,
        selectedTab: tabIndex
    };
};
const compareTabData = (tab1, tab2) => tab1.sectionId === tab2.sectionId && tab1.fileId === tab2.fileId;
const getTabIndex = (tabs, fileData) => tabs.findIndex(tab => compareTabData(fileData, tab));
const isTabExist = (tabs, fileData) => tabs.length !== 0 && tabs.findIndex(tab => compareTabData(fileData, tab)) !== -1;

const actionHandlers = {
    [STATE.TAB_ADD]: (state, fileData) => {
        let {tabs} = state;

        if (isTabExist(tabs, fileData)) {
            const tabIndex = getTabIndex(tabs, fileData);
            return selectTab(state, tabIndex);
        }

        return {
            ...state,
            tabs: [...tabs, fileData],
            selectedTab: tabs.length
        }
    },
    [STATE.TAB_SELECT]: selectTab,
    [STATE.TAB_CLOSE]: closeTab,
    [STATE.ZOOM_CHANGE]: (state, zoomValue) => ({...state, zoomValue}),
    [STATE.TRANSFORM_RESET]: state => ({...state, zoomValue: 1}),
    [STATE.CHECK_RENAME]: (state, fileData) => {
        const {tabs} = state;

        if (!isTabExist(tabs, fileData)) {
            return state;
        }

        return {
            ...state,
            tabs: tabs.map(tab => !compareTabData(tab, fileData) ?
                tab : {
                    ...tab,
                    title: fileData.title
                }
            )
        }
    },
    [STATE.CHECK_DELETE]: (state, fileData) => {
        const {tabs} = state;

        if (!isTabExist(tabs, fileData)) {
            return state;
        }

        const tabIndex = getTabIndex(tabs, fileData);

        return closeTab(state, tabIndex);
    }
};

export default function workingAreaReducer(state = initialState, {type, payload}) {
    const actionHandler = actionHandlers[type];
    return actionHandler ? actionHandler(state, payload) : state;
}
