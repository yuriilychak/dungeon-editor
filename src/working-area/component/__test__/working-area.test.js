import React from "react";

import { createShallow } from "@material-ui/core/test-utils";

import reducer from "../../reducer";
import WorkingArea from "../working-area";
import { getInitialState } from "../../../../test_templates";
import { UI_SECTION } from "../../../enum";

jest.mock("../../../working-stage");

describe("working-area test", () => {
    const initialState = getInitialState(reducer, UI_SECTION.WORKING_AREA);

    it ( 'working-area default snapshot', () => {
        const wrapper = createShallow()(
            <WorkingArea
                {...initialState}
                onCreateElement={jest.fn()}
                onGetCanvasRef={jest.fn()}
                onCloseTab={jest.fn()}
                onSelectTab={jest.fn()}
                onTransformReset={jest.fn()}
            />
        );

        expect(wrapper.html()).toMatchSnapshot();
    });

    it ( 'working-area snapshot with tabs', () => {
        const wrapper = createShallow()(
            <WorkingArea
                {...initialState}
                onGetCanvasRef={jest.fn()}
                onCloseTab={jest.fn()}
                onSelectTab={jest.fn()}
                onCreateElement={jest.fn()}
                onTransformReset={jest.fn()}
                tabs={[
                    {
                        title: "tab_1",
                        sectionId: 0,
                        fileId: 0
                    },
                    {
                        title: "tab_2",
                        sectionId: 1,
                        fileId: 1
                    }
                ]}
            />
        );

        expect(wrapper.html()).toMatchSnapshot();
    });
});
