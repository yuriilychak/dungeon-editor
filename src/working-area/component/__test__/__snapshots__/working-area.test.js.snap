// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`working-area test working-area default snapshot 1`] = `"<div class=\\"working-area-root\\"><div class=\\"working-area-header\\"><div class=\\"MuiTabs-root makeStyles-root-2\\"><div class=\\"MuiTabs-scroller MuiTabs-fixed\\" style=\\"overflow:hidden\\"><div class=\\"MuiTabs-flexContainer\\" role=\\"tablist\\"><button class=\\"MuiButtonBase-root MuiTab-root makeStyles-root-1 MuiTab-textColorInherit Mui-selected\\" tabindex=\\"0\\" type=\\"button\\" role=\\"tab\\" aria-selected=\\"true\\"><span class=\\"MuiTab-wrapper\\"><div class=\\"tab-content-root\\"><img src=\\"undefined/static/icon/directory.svg\\" style=\\"width:10px;height:10px\\" class=\\"icon-root\\" alt=\\"empty Icon\\" draggable=\\"false\\"/><span class=\\"tab-content-title\\">Default</span></div></span><span class=\\"PrivateTabIndicator-root-25 PrivateTabIndicator-colorSecondary-27 MuiTabs-indicator\\"></span></button></div></div></div></div><div class=\\"working-area-body\\"><div class=\\"working-area-interactions\\"><canvas id=\\"WorkingCanvas\\" class=\\"working-canvas-root working-canvas-hidden\\"></canvas></div><div class=\\"working-area-empty-container\\"><div class=\\"working-area-empty-message\\">For open some item from library make double click on it.</div></div></div></div>"`;

exports[`working-area test working-area snapshot with tabs 1`] = `"<div class=\\"working-area-root\\"><div class=\\"working-area-header\\"><div class=\\"MuiTabs-root makeStyles-root-2\\"><div class=\\"MuiTabs-scroller MuiTabs-fixed\\" style=\\"overflow:hidden\\"><div class=\\"MuiTabs-flexContainer\\" role=\\"tablist\\"><button class=\\"MuiButtonBase-root MuiTab-root makeStyles-root-1 MuiTab-textColorInherit Mui-selected\\" tabindex=\\"0\\" type=\\"button\\" role=\\"tab\\" aria-selected=\\"true\\"><span class=\\"MuiTab-wrapper\\"><div class=\\"tab-content-root\\"><img src=\\"undefined/static/icon/element_element.svg\\" style=\\"width:10px;height:10px\\" class=\\"icon-root\\" alt=\\"empty Icon\\" draggable=\\"false\\"/><span class=\\"tab-content-title\\">tab_1</span><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"></path><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path></svg></div></span><span class=\\"PrivateTabIndicator-root-25 PrivateTabIndicator-colorSecondary-27 MuiTabs-indicator\\"></span></button><button class=\\"MuiButtonBase-root MuiTab-root makeStyles-root-1 MuiTab-textColorInherit\\" tabindex=\\"0\\" type=\\"button\\" role=\\"tab\\" aria-selected=\\"false\\"><span class=\\"MuiTab-wrapper\\"><div class=\\"tab-content-root\\"><img src=\\"undefined/static/icon/font_element.svg\\" style=\\"width:10px;height:10px\\" class=\\"icon-root\\" alt=\\"empty Icon\\" draggable=\\"false\\"/><span class=\\"tab-content-title\\">tab_2</span><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"></path><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path></svg></div></span></button></div></div></div></div><div class=\\"working-area-body\\"><div class=\\"working-area-interactions\\"><canvas id=\\"WorkingCanvas\\" class=\\"working-canvas-root\\"></canvas></div><div class=\\"working-area-control-panel\\"><button class=\\"MuiButtonBase-root MuiIconButton-root\\" tabindex=\\"0\\" type=\\"button\\" aria-haspopup=\\"false\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M19 8l-4 4h3c0 3.31-2.69 6-6 6-1.01 0-1.97-.25-2.8-.7l-1.46 1.46C8.97 19.54 10.43 20 12 20c4.42 0 8-3.58 8-8h3l-4-4zM6 12c0-3.31 2.69-6 6-6 1.01 0 1.97.25 2.8.7l1.46-1.46C15.03 4.46 13.57 4 12 4c-4.42 0-8 3.58-8 8H1l4 4 4-4H6z\\"></path><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path></svg></span></button><span class=\\"MuiSlider-root MuiSlider-colorPrimary MuiSlider-marked\\"><span class=\\"MuiSlider-rail\\"></span><span class=\\"MuiSlider-track\\" style=\\"left:0%;width:20%\\"></span><input type=\\"hidden\\" value=\\"1\\"/><span style=\\"left:0%\\" class=\\"MuiSlider-mark MuiSlider-markActive\\"></span><span aria-hidden=\\"true\\" style=\\"left:0%\\" class=\\"MuiSlider-markLabel MuiSlider-markLabelActive\\">25%</span><span style=\\"left:20%\\" class=\\"MuiSlider-mark MuiSlider-markActive\\"></span><span aria-hidden=\\"true\\" style=\\"left:20%\\" class=\\"MuiSlider-markLabel MuiSlider-markLabelActive\\">100%</span><span style=\\"left:46.666666666666664%\\" class=\\"MuiSlider-mark\\"></span><span aria-hidden=\\"true\\" style=\\"left:46.666666666666664%\\" class=\\"MuiSlider-markLabel\\">200%</span><span style=\\"left:73.33333333333333%\\" class=\\"MuiSlider-mark\\"></span><span aria-hidden=\\"true\\" style=\\"left:73.33333333333333%\\" class=\\"MuiSlider-markLabel\\">300%</span><span style=\\"left:100%\\" class=\\"MuiSlider-mark\\"></span><span aria-hidden=\\"true\\" style=\\"left:100%\\" class=\\"MuiSlider-markLabel\\">400%</span><span class=\\"MuiSlider-thumb MuiSlider-thumbColorPrimary PrivateValueLabel-thumb-67\\" tabindex=\\"0\\" role=\\"slider\\" style=\\"left:20%\\" data-index=\\"0\\" aria-labelledby=\\"discrete-slider\\" aria-orientation=\\"horizontal\\" aria-valuemax=\\"4\\" aria-valuemin=\\"0.25\\" aria-valuenow=\\"1\\" aria-valuetext=\\"100%\\"><span class=\\"PrivateValueLabel-offset-69 MuiSlider-valueLabel\\"><span class=\\"PrivateValueLabel-circle-70\\"><span class=\\"PrivateValueLabel-label-71\\">100%</span></span></span></span></span><div class=\\"working-area-control-panel-mode\\">Size</div></div></div></div>"`;
